scripts-dev/mypy_synapse_plugin.py

```diff
@@ -92,10 +93,35 @@ def cached_function_method_signature(ctx: MethodSigContext) -> CallableType:
     arg_names.append("on_invalidate")
     arg_kinds.append(ARG_NAMED_OPT)  # Arg is an optional kwarg.
 
+    # Finally we ensure the return type is a Deferred.
+    if (
+        isinstance(signature.ret_type, Instance)
+        and signature.ret_type.type.fullname == "twisted.internet.defer.Deferred"
+    ):
+        # If it is already a Deferred, nothing to do.
+        ret_type = signature.ret_type
+    else:
+        # If a coroutine, wrap the coroutine's return type in a Deferred.
+        if (
+            isinstance(signature.ret_type, Instance)
+            and signature.ret_type.type.fullname == "typing.Coroutine"
```

Comment:
These cases don't cover the cached function returning a generic `Awaitable[R]` . I don't _think_ we use Awaitables that aren't Deferreds or Coroutines in Synapse's code (outside of tests, at least), so this is probably fine. We can come back to this in the future if need be.

Reply:
Yeah, we don't seem to do that so I didn't bother covering it. I should likely just add it while I'm here though.